lib/concurrent-merkle-tree/src/state.rs

```diff
@@ -23,27 +23,46 @@ impl<const MAX_DEPTH: usize> ChangeLog<MAX_DEPTH> {
         }
     }
 
+    pub fn new(root: Node, path: [Node; MAX_DEPTH], index: u32) -> Self {
+        Self {
+            root,
+            path,
+            index,
+            _padding: 0,
+        }
+    }
+
     pub fn get_leaf(&self) -> Node {
         self.path[0]
     }
 
     /// Sets all change log values from a leaf and valid proof
-    pub fn recompute_path(&mut self, mut start: Node, proof: &[Node]) -> Node {
-        self.path[0] = start;
-        for (ix, s) in proof.iter().enumerate() {
-            if self.index >> ix & 1 == 0 {
-                let res = hashv(&[start.as_ref(), s.as_ref()]);
-                start.copy_from_slice(res.as_ref());
-            } else {
-                let res = hashv(&[s.as_ref(), start.as_ref()]);
-                start.copy_from_slice(res.as_ref());
-            }
-            if ix < MAX_DEPTH - 1 {
-                self.path[ix + 1] = start;
-            }
+    pub fn recompute_path(&mut self, index: u32, mut node: Node, proof: &[Node]) -> Node {
+        self.index = index;
```

Comment:
why does recomputing path set the `index` of the leaf?

Reply:
you're updating the change log